File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    let max = 0;
    
    function maxDepth(root: TreeNode | null) {
        if(!root) return 0;
        let left = maxDepth(root.left);
        let right = maxDepth(root.right);
        //compare old max, with new max, where new max,
        max = Math.max(max, left + right);
        return Math.max(left, right) + 1;
    }
    maxDepth(root);
    return max;
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
            self.ans = 0
    
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def maxDepth(root):
            if not root: return 0
            left = maxDepth(root.left)
            right = maxDepth(root.right)
            new_ans = left + right
            self.ans = max(self.ans, new_ans)
            return max(left, right) + 1
        maxDepth(root)
        return self.ans
        